[GLIB] Add more signpost marks related to graphics

GeorgesStavracas opened the pull request:
#### f6853da27ea8bf30133d2c1b6cdc3265ec75b37c
<pre>
[GLIB] Add more signpost marks related to graphics
<a href="https://bugs.webkit.org/show_bug.cgi?id=278411">https://bugs.webkit.org/show_bug.cgi?id=278411</a>

Reviewed by Nikolas Zimmermann.

Add a series of marks that act as submarks of BackingStoreFlush. In
particular, the important ones are the marks tile renders.

* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/glib/SysprofAnnotator.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::commitSceneState):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::finalizeRenderingUpdate):

Canonical link: <a href="https://commits.webkit.org/282779@main">https://commits.webkit.org/282779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/29893bd91328176135e83f8384b8f3cf54155a31

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64197 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43554 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16794 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14805 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66317 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/51678 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10215 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67266 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40279 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55549 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32297 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36949 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12932 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13679 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/57309 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58911 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13261 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69918 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/63442 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8144 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12781 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/58996 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8177 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55645 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59155 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14180 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6736 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/424 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85203 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39374 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15026 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40453 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41636 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40196 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WTF/wtf/SystemTracing.h:
```diff
@@ -150,6 +150,8 @@ enum TracePointCode {
     SyntheticMomentumEvent,
     RemoteLayerTreeScheduleRenderingUpdate,
     DisplayLinkUpdate,
+    UpdateLayerContentBuffersStart,
```
These shouldn't be here but in the `GTKWPEPortRange` below, otherwise you need to update the `Tools/Tracing/SystemTracePoints.plist` file, to make sure it won't break the Apple ports, once they add another enum entry below this one. So better move to our section.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
```diff
@@ -1221,6 +1224,8 @@ void CoordinatedGraphicsLayer::updateContentBuffers()
     if (!dirtyTiles.isEmpty()) {
         bool didUpdateTiles = false;
 
+        WTFBeginSignpost(this, PaintTiles, "dirty tiles: %lu", dirtyTiles.size());
```
I would also add signposts in the inner loop over the dirtyTiles including information such as which tiles is updated: tile 1/N, tileRect + dirtyRect + tileID -- this allows for visual inspection when tiles first appear, are re-used or disappear. Also helpful to track the dirtyRect in sysprof.

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -81,7 +82,10 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
     if (auto* acceleratedBufferPool = m_coordinator->skiaAcceleratedBufferPool()) {
         PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
         if (auto buffer = acceleratedBufferPool->acquireBuffer(tileRect.size(), !contentsOpaque())) {
+            WTFBeginSignpost(this, PaintTile, "Skia accelerated");
```
How about including the buffer size?

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -91,6 +95,8 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
 
     // Non-blocking, multi-threaded variant.
     if (auto* workerPool = m_coordinator->skiaUnacceleratedThreadedRenderingPool()) {
+        WTFBeginSignpost(this, PaintTile, "Skia unaccelerated multithread");
```
This is useful, but even more useful would be to measure independently the time for the recording of the display list, and then in the postTask() lambda, the time for the replay + paint operation (ideally including the thread ID, so we can distinguish which thread was doing work -- or does sysprof already incldue that in the mark table, can't recall right now).

nikolaszimmermann replied:
@carlosgcampos 
I already suggested that and thought it was done.

GeorgesStavracas replied:
Done for real now. Sorry about this!

nikolaszimmermann commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -106,11 +112,16 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
     }
 
     // Blocking, single-thread variant.
+    WTFBeginSignpost(this, PaintTile, "Skia unaccelerated");
```
Include the buffer size in the description.

nikolaszimmermann commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
```diff
@@ -170,20 +170,26 @@ bool CompositingCoordinator::flushPendingLayerChanges(OptionSet<FinalizeRenderin
     m_page.updateRendering();
     m_page.flushPendingEditorStateUpdate();
 
+    WTFBeginSignpost(this, FlushRootCompositingLayer);
```
I'm not a compositing expert - maybe @magomez could suggest other interessting things to include here. Looks good to me.

nikolaszimmermann reviewed (commented):
Great stuff -- I've been working on similar signpost messages to check the dirtyRects in the tiling code in sysprof -- great opportunity to combine the effort.

nikolaszimmermann reviewed (approved):
r=me

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
```diff
@@ -91,6 +95,8 @@ Ref<Nicosia::Buffer> CoordinatedGraphicsLayer::paintTile(const IntRect& tileRect
     // Non-blocking, multi-threaded variant.
     if (auto* workerPool = m_coordinator->skiaUnacceleratedThreadedRenderingPool()) {
+        WTFBeginSignpost(this, PaintTile, "Skia unaccelerated multithread, dirty region %ix%i+%i+%i", tileRect.x(), tileRect.y(), tileRect.width(), tileRect.height());
+
```
This is not the time to paint the tile, but the time to record the painting commands. Maybe we should rename it as RecordTile and add the Paint tile in the worker thread before and after the display list is replayed.

nikolaszimmermann replied:
Good catch.

GeorgesStavracas replied:
Done!

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
```diff
@@ -369,8 +369,12 @@ void LayerTreeHost::didFlushRootLayer(const FloatRect& visibleContentRect)
     m_isWaitingForRenderer = true;
     m_compositor->updateSceneState(state, ++m_compositionRequestID);
+
+    WTFEndSignpost(this, LayerTreeHostCommitScene);
```
I'm not sure it's useful to know the duration of updateSceneState, it's just adding an element to a Vector and scheduling a run loop source. It should always be fast.

GeorgesStavracas replied:
Removed

nikolaszimmermann reviewed (approved):
Let's get this merged 🥳

GeorgesStavracas committed:
[GLIB] Add more signpost marks related to graphics
https://bugs.webkit.org/show_bug.cgi?id=278411

Reviewed by Nikolas Zimmermann.

Add a series of marks that act as submarks of BackingStoreFlush. In
particular, the important ones are the marks tile renders.

* Source/WTF/wtf/SystemTracing.h:
* Source/WTF/wtf/glib/SysprofAnnotator.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::updateContentBuffers):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayerSkia.cpp:
(WebCore::CoordinatedGraphicsLayer::paintTile):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/LayerTreeHost.cpp:
(WebKit::LayerTreeHost::commitSceneState):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::finalizeRenderingUpdate):

Canonical link: https://commits.webkit.org/282779@main
